File: algorithms/algorithms-cplusplus/leetcode/Question_0145_Binary_Tree_Postorder_Traversal.cc
#include <vector>
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<int> postorderTraversal(TreeNode* root) {
    vector<int> res;
    if (!root) {
      return res;
    }
    stack<TreeNode*> s;
    TreeNode* last = nullptr;
    while (root || !s.empty()) {
      if (root) {
        s.push(root);
        root = root->left;
      } else {
        TreeNode* top = s.top();
        if (top->right && top->right != last) {
          root = top->right;
        } else {
          res.push_back(top->val);
          last = top;
          s.pop();
        }
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0145_Binary_Tree_Postorder_Traversal.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Question_0145_Binary_Tree_Postorder_Traversal {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        TreeNode last = null;
        Stack<TreeNode> stack = new Stack<>();
        while (root != null || !stack.empty()) {
            if (root != null) {
                stack.push(root);
                root = root.left;
            } else {
                TreeNode node = stack.peek();
                if (node.right != null && node.right != last) {
                    root = node.right;
                } else {
                    stack.pop();
                    res.add(node.val);
                    last = node;
                }
            }
        }
        return res;
    }
}
